### PR TITLE
Fix for bug #922 & bug #949 

### PR DESCRIPTION
**Pull request by HugoDoyon**
Fix for "Do not propose lightning payment if value of the invoice is above..." and minimum on chain amount. For your review ;)

### CODE REVIEW

**Comment by NicolasDorier** [comment]
I am not super fan of such whitebox testing and would prefer a test via selenium or through the real server. But that is not blocking, @pavlenex you can test it?

**Comment by NicolasDorier** [comment]
This is not correct, the OnChainMinValue should remove onchain payments.

File: BTCPayServer/Payments/Bitcoin/BitcoinLikePaymentHandler.cs [hunk 1]
```diff
@@ -73,37 +69,62 @@ public override string GetPaymentMethodName(PaymentMethodId paymentMethodId)
...
             Dictionary<CurrencyPair, Task<RateResult>> rate, Money amount, PaymentMethodId paymentMethodId)
         {
+            
             Func<Money, Money, bool> compare = null;
             CurrencyValue limitValue = null;
             string errorMessage = null;
-            if (paymentMethodId.PaymentType == PaymentTypes.LightningLike &&
-               storeBlob.LightningMaxValue != null)
+            
+            if (paymentMethodId.PaymentType == PaymentTypes.BTCLike)
```
Comment (NicolasDorier) [inline BTCPayServer/Payments/Bitcoin/BitcoinLikePaymentHandler.cs]: remove this test, it is actually useless, we are always BTCLike if we reach here.

File: BTCPayServer/Payments/Bitcoin/BitcoinLikePaymentHandler.cs [hunk 2]
```diff
@@ -73,37 +69,62 @@ public override string GetPaymentMethodName(PaymentMethodId paymentMethodId)
...
-                limitValue = storeBlob.LightningMaxValue;
-                errorMessage = "The amount of the invoice is too high to be paid with lightning";
-            }
-            else if (paymentMethodId.PaymentType == PaymentTypes.BTCLike &&
-               storeBlob.OnChainMinValue != null)
-            {
-                compare = (a, b) => a < b;
-                limitValue = storeBlob.OnChainMinValue;
-                errorMessage = "The amount of the invoice is too low to be paid on chain";
+                if (storeBlob.LightningMaxValue == null &&
```
Comment (NicolasDorier) [inline BTCPayServer/Payments/Bitcoin/BitcoinLikePaymentHandler.cs]: `LightningMaxValue ` is orthogonal to onchain payment, no test should be done on that.

File: BTCPayServer/Payments/Bitcoin/BitcoinLikePaymentHandler.cs [hunk 3]
```diff
@@ -73,37 +69,62 @@ public override string GetPaymentMethodName(PaymentMethodId paymentMethodId)
...
-                limitValue = storeBlob.OnChainMinValue;
-                errorMessage = "The amount of the invoice is too low to be paid on chain";
+                if (storeBlob.LightningMaxValue == null && 
+                    storeBlob.OnChainMinValue != null)
+                {
+                    compare = (value, limit) => value < limit;
+                    limitValue = storeBlob.OnChainMinValue;
+                    errorMessage = "The amount of the invoice is too low to be paid on chain";
+                }
+                else if (storeBlob.LightningMaxValue != null &&
```
Comment (NicolasDorier) [inline BTCPayServer/Payments/Bitcoin/BitcoinLikePaymentHandler.cs]: same

**Comment by NicolasDorier** [comment]
I made the code more complicated that it should have been. 
Basically the lightning handler should only use MaxLightningValue while the bitcoin handler should only use MinLightningValue.

No need for the strange hacks I did.

File: BTCPayServer/Payments/Bitcoin/BitcoinLikePaymentHandler.cs (hunk 1 above)
Comment (HugoDoyon) [reply BTCPayServer/Payments/Bitcoin/BitcoinLikePaymentHandler.cs]: No problem, that actually overkill indeed

File: BTCPayServer/Payments/Bitcoin/BitcoinLikePaymentHandler.cs (hunk 2 above)
Comment (HugoDoyon) [reply BTCPayServer/Payments/Bitcoin/BitcoinLikePaymentHandler.cs]: That is true LightningMaxValue and OnChainMinValue are orthogonal. 

However, while testing I realize that a strange edge case could happen if LightningMaxValue is smaller than OnChainMinValue. In that situation there will be a gap where the no payment method will be available to the user. 

For exemple: IF LightningMaxValue = 50 USD$ AND OnChainMinValue = 52 USD$ AND InvoiceAmount =51 USD$ THEN No payment method available for user

**Comment by HugoDoyon** [comment]
Glad to know that you guys prefer E2E test (Selenium/CodeUI/Protractor) over simple unit test as I do like them more too and I also very comfortable with those tech as well. Next time, I will propose those kind of test.

File: BTCPayServer/Payments/Bitcoin/BitcoinLikePaymentHandler.cs (hunk 2 above)
Comment (NicolasDorier) [reply BTCPayServer/Payments/Bitcoin/BitcoinLikePaymentHandler.cs]: That's fine. The use will get an error which explain clearly the issue.

**Comment by HugoDoyon** [comment]
Change done ;)

**Pull request merged**